applications/OptimizationApplication/custom_utilities/geometrical/symmetry_utility.cpp

```diff
@@ -0,0 +1,270 @@
+{
+
+    if(mSymmetrySettings["type"].GetString()=="plane_symmetry"){
+        mPlaneSymmetry = true;
+        mPlaneSymmetryData.Point = mSymmetrySettings["settings"]["point"].GetVector();
+        array_3d normal = mSymmetrySettings["settings"]["normal"].GetVector();
+        const double norm_normal = norm_2(normal);
+        if (norm_normal < std::numeric_limits<double>::epsilon())
+            KRATOS_ERROR<<"SymmetryUtility: norm of normal is close to zero"<<std::endl;
+            
+        mPlaneSymmetryData.Normal = normal/norm_normal;
+        mPlaneSymmetryData.ReflectionMatrix = IdentityMatrix(3) - (2*outer_prod(mPlaneSymmetryData.Normal, mPlaneSymmetryData.Normal));
+    }
+    else if(mSymmetrySettings["type"].GetString()=="rotational_symmetry"){
+        mAxisSymmetry = true;
+        mRotationalSymmetryData.Point = mSymmetrySettings["settings"]["point"].GetVector();
+        array_3d axis = mSymmetrySettings["settings"]["axis"].GetVector();
+        const double norm_axis = norm_2(axis);
+        if (norm_axis < std::numeric_limits<double>::epsilon())
+            KRATOS_ERROR<<"SymmetryUtility: norm of axis is close to zero"<<std::endl;
+            
+        mRotationalSymmetryData.Axis = axis/norm_axis;
+        mRotationalSymmetryData.Angle = mSymmetrySettings["settings"]["angle"].GetDouble();        
+    }
+    else
+        KRATOS_ERROR<<"SymmetryUtility: type should be either plane or axis"<<std::endl;
+
+}
+
+
+
+void SymmetryUtility::Initialize()
+{
+
+    BuiltinTimer timer;
+    KRATOS_INFO("SymmetryUtility::Initialize: ") << "Starting initialization of symmetry utility "<<mUtilName<<" ..." << std::endl;
+
+    NodeVector mListOfNodesInModelPart;
+    mListOfNodesInModelPart.resize(mrModelPart.Nodes().size());
+    int counter = 0;
+    for (ModelPart::NodesContainerType::iterator node_it = mrModelPart.NodesBegin(); node_it != mrModelPart.NodesEnd(); ++node_it)
+    {
+        NodeTypePointer pnode = *(node_it.base());
+        mListOfNodesInModelPart[counter++] = pnode;
+    }
+
+    const size_t bucket_size = 100;
+    KDTree search_tree(mListOfNodesInModelPart.begin(), mListOfNodesInModelPart.end(), bucket_size);
+
+    if(mAxisSymmetry){
+
+        #pragma omp declare reduction (merge : std::vector<std::pair <NodeTypePointer,NodeVector>> : omp_out.insert(omp_out.end(), omp_in.begin(), omp_in.end()))        
+        std::vector<std::pair <NodeTypePointer,NodeVector>>& rMap = mRotationalSymmetryData.Map;
+        #pragma omp parallel for reduction(merge: rMap)
+        for (auto& r_node : mrModelPart.Nodes()){
+            int num_rotations = (360.0/mRotationalSymmetryData.Angle);
+            NodeVector p_neighbor_rot_nodes;
+            p_neighbor_rot_nodes.resize(num_rotations-1);
+            for(int r_i=1;r_i<num_rotations;r_i++){
+               NodeTypePointer p_rot_node = GetRotatedNode(r_node,r_i*mRotationalSymmetryData.Angle);
```

Comment:
this is extremely expensive; you're allocating and computing the rotation matrix multiple times for every single node. Consider computing the transforms before looping through the nodes.

Reply:
Although I would say computing 3by3 matrix is not be expensive, I will precompute the matrices and save to be used later.